lib/eval.ml

```diff
@@ -1241,19 +1399,23 @@ module MakeInterpreter (T : Target) = struct
   (*----------------------------------------------------------------------------*)
 
   and eval_parser (ctrl : ctrl) (env : env) (st : state) (params : TypeParameter.t list)
-      (args : Expression.t option list) (vs : (string * value) list)
+      (args : Expression.t option list) (pscope : env) (ls : (string * loc) list)
       (locals : Declaration.t list) (states : Parser.state list) : env * state * signal =
-    let (penv, st', s) = copyin ctrl env st params args in
+    (* TODO: incorporate closure environment *)
+    let (callenv,penv, st', s) = copyin ctrl env st env params args in
     match s with
     | SContinue ->
       let f a (x,y) = EvalEnv.insert_val_bare x y a in
-      let penv' = List.fold_left vs ~init:penv ~f:f in
+      let penv' = List.fold ls ~init:penv ~f:f in
       let (penv'', st'') = List.fold_left locals ~init:(penv',st') ~f:(fun (e,s) -> eval_decl ctrl e s) in
       let states' = List.map states ~f:(fun s -> snd (snd s).name, s) in
       let start = find_exn states' "start" in
       let (penv''', st''', final_state) = eval_state_machine ctrl penv'' st'' states' start in
-      (copyout ctrl penv''' st''' params args false, st''', final_state)
-    | SReject _ -> (EvalEnv.pop_scope penv, st, s)
+      (* print_endline "copy out from parser"; *)
+      let st'''' = copyout ctrl callenv penv''' st''' params args false in
```

Comment:
It's better to shadow names sometimes so you don't accumulate so many prime marks `'''`

Reply:
cleaned up a few instances of it